src/main/java/com/ontotext/kafka/processor/SinkRecordsProcessor.java

```diff
@@ -0,0 +1,320 @@
+		MDC.put("RepositoryURL", repositoryUrl);
+		MDC.put("Connector", config.getConnectorName());
+	}
+
+
+	private static Repository initRepository(GraphDBSinkConfig config) {
+		String address = config.getServerUrl();
+		String repositoryId = config.getRepositoryId();
+		GraphDBSinkConfig.AuthenticationType authType = config.getAuthType();
+		HTTPRepository repository = new HTTPRepository(address, repositoryId);
+		switch (authType) {
+			case NONE:
+				return repository;
+			case BASIC:
+				if (LOG.isTraceEnabled()) {
+					LOG.trace("Initializing repository connection with user {}", config.getAuthBasicUser());
+				}
+				repository.setUsernameAndPassword(config.getAuthBasicUser(), config.getAuthBasicPassword().value());
+				return repository;
+			case CUSTOM:
+			default: // Any other types which are valid, as per definition, but are not implemented yet
+				throw new UnsupportedOperationException(authType + " not supported");
+		}
+	}
+
+	@Override
+	public void run() {
+		while (!Thread.currentThread().isInterrupted()) {
+			try {
+				if (backOff.getAndSet(false)) {
+					LOG.info("Retrying flush");
+					flushUpdates();
+				}
+				Collection<SinkRecord> messages = sinkRecords.poll(timeoutCommitMs, TimeUnit.MILLISECONDS);
+				if (messages != null) {
+					consumeRecords(messages);
+				} else {
+					LOG.trace("Did not receive any records (waited {} {}} ) for repository {}. Flushing all records in batch.", timeoutCommitMs,
+						TimeUnit.MILLISECONDS, repository);
+					flushUpdates();
+				}
+			} catch (RetriableException e) {
+				long backoffTimeoutMs = config.getBackOffTimeoutMs();
+				backOff.set(true);
+				LOG.warn("Caught a retriable exception while flushing the current batch. " +
+					"Will sleep for {}ms and will try to flush the records again", backoffTimeoutMs);
+				try {
+					Thread.sleep(backoffTimeoutMs);
+				} catch (InterruptedException iex) {
+					LOG.info("Thread was interrupted during backoff. Shutting down");
+					break;
+				}
+
+
+			} catch (Exception e) {
+				if (e instanceof InterruptedException) {
+					LOG.info("Thread was interrupted. Shutting down processor");
+				} else {
+					LOG.error("Caught an exception, cannot recover.  Shutting down", e);
+				}
+				break;
+			}
+		}
+		Thread.interrupted();
+		shutdown();
+	}
+
+	private void shutdown() {
+		// commit any records left before shutdown. Bypass batch checking, just add all records and flush downstream
+		LOG.info("Commiting any records left before shutdown");
+		Collection<SinkRecord> records;
+		while ((records = sinkRecords.poll()) != null) {
+
+			recordsBatch.addAll(records);
+		}
+		try {
+			flushUpdates();
+		} catch (Exception e) {
+			// We did ou best. Just log the exception and shut down
+			LOG.warn("While shutting down, failed to flush updates due to exception", e);
+		}
+
+		if (repository.isInitialized()) {
```

Comment:
If we move out the repository initialisation outside, perhaps we can move the shutdown too?